Source/inv.cpp

```diff
@@ -347,50 +491,89 @@ void CheckInvPaste(Player &player, Point cursorPosition)
 
-	item_equip_type desiredIl = player.GetItemLocation(player.HoldItem);
-	if (il == ILOC_ONEHAND && desiredIl == ILOC_TWOHAND)
-		il = ILOC_TWOHAND;
-
-	int8_t it = 0;
-	if (il == ILOC_UNEQUIPABLE) {
-		int ii = slot - SLOTXY_INV_FIRST;
-		if (player.HoldItem._itype == ItemType::Gold) {
-			if (player.InvGrid[ii] != 0) {
-				int8_t iv = player.InvGrid[ii];
-				if (iv > 0) {
-					if (player.InvList[iv - 1]._itype != ItemType::Gold) {
-						it = iv;
+	if (il == ILOC_ONEHAND && desired_loc == ILOC_TWOHAND)
+		return ILOC_TWOHAND;
+
+	return il;
+}
+
+std::optional<int8_t> CheckOverlappingItems(int slot, const Player &player, Size itemSize, int8_t prev_it)
+{
+	// check that the item we're pasting only overlaps one other item (or is going into empty space)
+	const unsigned originCell = static_cast<unsigned>(slot - SLOTXY_INV_FIRST);
+
+	int8_t it = prev_it;
+	for (unsigned rowOffset = 0; rowOffset < static_cast<unsigned>(itemSize.height * InventorySizeInSlots.width); rowOffset += InventorySizeInSlots.width) {
+
+		for (unsigned columnOffset = 0; columnOffset < static_cast<unsigned>(itemSize.width); columnOffset++) {
+			unsigned testCell = originCell + rowOffset + columnOffset;
+			// FindTargetSlotUnderItemCursor returns the top left slot of the inventory region that fits the item, we can be confident this calculation is not going to read out of range.
+			assert(testCell < sizeof(player.InvGrid));
+			if (player.InvGrid[testCell] != 0) {
+				int8_t iv = std::abs(player.InvGrid[testCell]);
+				if (it != 0) {
+					if (it != iv) {
+						// Found two different items that would be displaced by the held item, can't paste the item here.
+						return std::nullopt;
 					}
 				} else {
-					it = -iv;
+					it = iv;
 				}
 			}
-		} else {
-			// check that the item we're pasting only overlaps one other item (or is going into empty space)
-			unsigned originCell = static_cast<unsigned>(slot - SLOTXY_INV_FIRST);
-			for (unsigned rowOffset = 0; rowOffset < static_cast<unsigned>(itemSize.height * InventorySizeInSlots.width); rowOffset += InventorySizeInSlots.width) {
-				for (unsigned columnOffset = 0; columnOffset < static_cast<unsigned>(itemSize.width); columnOffset++) {
-					unsigned testCell = originCell + rowOffset + columnOffset;
-					// FindTargetSlotUnderItemCursor returns the top left slot of the inventory region that fits the item, we can be confident this calculation is not going to read out of range.
-					assert(testCell < sizeof(player.InvGrid));
-					if (player.InvGrid[testCell] != 0) {
-						int8_t iv = std::abs(player.InvGrid[testCell]);
-						if (it != 0) {
-							if (it != iv) {
-								// Found two different items that would be displaced by the held item, can't paste the item here.
-								return;
-							}
-						} else {
-							it = iv;
-						}
-					}
+		}
+	}
+
+	return it;
+}
+
+std::optional<int8_t> GetPrevItemId(int slot, const Player &player, const Size &itemSize)
+{
+	int8_t item_something = 0;

```

Comment:
Perhaps a better variable name? This is quite ambigious

Reply:
I agree, hence why this feedback request! I guess it gets the item ID from the inv grid? I was not too sure on this one.

Reply:
Looks like `prevItemId` would have been a good name since that reflects the getter function name and this is the returned value.